### ISSUE TITLE
AST struct transfer, serialization, deserialization performance enhancement

### ISSUE THREAD

**Issue by overlookmotel** [swc-project/swc#2175]
As mentioned in #2123 and #1392, `parse()` is currently quite slow.

SWC's Rust parser is of course extremely fast, but the bottleneck is passing ASTs back into JS.

I've done some experiments, and believe I may have the beginnings of a solution to speed it up by 5x or more, enough so it out-performs Babel at least.

https://github.com/overlookmotel/swc-parse-test

![1000 lines](https://raw.githubusercontent.com/overlookmotel/swc-parse-test/master/doc/images/1000_lines_brief.png)

### The problem

There are 2 bottlenecks involved in transferring AST from Rust to JS:

1. Cost of passing AST from Rust to JS (as JSON string)
2. Cost of parsing JSON in JS (using `JSON.parse()`)

### Potential solution

I've attempted to alleviate these pinch points by outputting a buffer (`JsBuffer`) from Rust instead of a JSON string. The buffer is a binary serialization of the AST, and I've written a matching deserializer in JS.

This has 2 advantages:

1. `JsBuffer` can be created in main thread's memory so passing it from Rust to JS is zero cost.
2. AST can be encoded and deserialized more efficiently in binary form.

The results appear to be pretty good.

### Provisos

The implementation is only a proof of concept. So far it can only deal with `const x = 1;` statements. I see no reason why it shouldn't maintain it's speed when support for all AST node types is included, but I could be wrong.

It may also be naive. Until this weekend, I've never written a line of Rust before, so I may be totally missing a reason why it's not possible to integrate this approach into SWC.

The Rust code I've written is, for the same reason, quite embarrassingly bad. The JS code is also hacked together very quickly. So potentially it could get significantly faster with some optimization.

### Questions

* Does this seem like a promising direction?

* If `parse()` (and `print()`) can be made much faster, might this make it feasible to maintain support for plugins written in Javascript?

As mentioned above, I don't know Rust at all, so I have limited ability to help with an effort to flesh out these experiments into fully-fledged features. However, I would be very happy to help out on the JS side (the deserializer etc).

**Comment by kdy1** [comment]
Yeah, if we can improve performance of passing ast, I expect the js plugin api to be really useful.

**Comment by Brooooooklyn** [comment]
Maybe deserialize via https://github.com/rkyv/rkyv in the wasm and convert into JsObject will be more faster.

**Comment by overlookmotel** [comment]
@Brooooooklyn rkyv looks great, in particular the zero-cost deserialization would be very efficient for getting an AST into `print()`.

But I'm afraid I don't understand your point about WASM. Can you possibly elaborate?

**Comment by Brooooooklyn** [comment]
From benchmark results before, the performance bottleneck in this scenario is `Create JsObject` in the non-JavaScript side.
And in my experience create `JsObject` in `wasm` is faster than in `Native` side.

**Comment by overlookmotel** [comment]
Thanks for quick reply? Are you suggesting that SWC would run as WASM in NodeJS, rather than via N-API? Or that you'd have a layer of WASM sitting in between JS and N-API?

Please excuse me if I have the terminology wrong. As I mentioned above, this is my first experience of Rust or Node native module development, so I am really just feeling my way here...

**Comment by overlookmotel** [comment]
And maybe this is a really stupid idea but...

Do we need to serialize at all? The AST already exists in Rust's memory in a binary format with a well-defined schema (the Rust types). Could that memory be wrapped in a JsBuffer (without copy), passed to JS, and a deserializer in JS would parse this buffer and convert it into JS objects?

The JS deserializer could be generated programmatically from the Rust type definitions.

For plugins, the deserializer could be lazy i.e. only deserialize sections of the buffer for AST nodes as and when are accessed in the plugin's visitors. For plugins which only operate on a subset of AST nodes, the serialization + deserialization overhead could be avoided entirely for the majority of the AST.

This sounds too good to be true. Am I totally misunderstanding what you can / can't do with data across the JS/Rust boundary?

**Comment by Brooooooklyn** [comment]
> For plugins, the deserializer could be lazy i.e. only deserialize sections of the buffer for AST nodes as and when they are accessed in the plugin's visitors.

simd_json has similar appraoch https://github.com/luizperes/simdjson_nodejs/issues/28

**Comment by kdy1** [comment]
> For plugins, the deserializer could be lazy i.e. only deserialize sections of the buffer for AST nodes as and when they are accessed in the plugin's visitors. For plugins which only operate on a subset of AST nodes, the serialization + deserialization overhead could be avoided entirely for the majority of the AST.

I expect this to slow down operations by a margin.
Visitors in plugins are expected to visit all nodes.

**Comment by kdy1** [comment]
I think we can use binary ser/de library like protobuf, which already has js-based deserializer.

**Comment by overlookmotel** [comment]
> Visitors in plugins are expected to visit all nodes.

Yes, there'd be no point in lazy-deserializing AST nodes if they're all going to be visited.

What I had in mind is to modify the implementation of the plugin system so it minimizes the number of nodes accessed.

For example, the [`ConsoleStripper` plugin example in SWC docs](https://swc.rs/docs/usage-plugin) only *needs* to visit a small fraction of nodes in an AST - `CallExpression`s and a fairly shallow subset of their child nodes.

I imagined it working something like this:

Plugins would be objects rather than functions (more like Babel I suppose):

```js
const consoleStripperPlugin = {
  visitor: {
    CallExpression(e: CallExpression): Expression {
      if (e.callee.type !== "MemberExpression") return e;
      // ... etc ...
      return e;
    }
  }
}
```

1. `transform()` examines the visitor and compiles an array of all the node types the visitor needs to visit initially (in this case just `['CallExpression']`).
2. `transform()` calls parse with the source JS + this array of node types.
3. `parse()` returns serialized AST (as buffer) + an array of pointers into that buffer for where all `CallExpression` nodes are located.
4. `transform()` deserializes all the `CallExpression` nodes to JS objects.
5. `transform()` calls `visitCallExpression()` with each of these node objects.
76 When visitor accesses `node. callee`, that AST node is lazily deserialized

The plugin can complete its work without accessing the majority of the AST, therefore there's no need to deserialize most of it.

Whatever the serialization format, deserialization in JS is going to be the slowest link in the chain, so it seems worthwhile to avoid as much of that work as possible.

I don't know if you'd be open to changing how plugins work in this direction? Implementation in SWC is more complex, but I imagine could be much more performant and, from a consumer point of view, probably a bit more familiar from Babel.

**Comment by overlookmotel** [comment]
@kdy1 I tried using serde to serialize to bincode instead of JSON, but was getting errors. I'll give protobuf a go.

Can anyone advise if my idea of using the binary representations of AST nodes in Rust's heap as the source for deserializing is (in theory at least) possible? Maybe it's a bad idea, but I'm curious as to whether it's in the realm of possibility, or whether it falls under "no, it doesn't work like that".

**Comment by Brooooooklyn** [comment]
Here is a similar discussion happened in Deno: https://github.com/denoland/deno/issues/2121

**Comment by yisar** [comment]
deno used similar techniques to optimize in the past, but their JSON structure is much simpler than the AST structure of SWC.
The tree of SWC is very deep, which will bring challenges.

**Comment by overlookmotel** [comment]
Thanks very much for this. Really interesting. I'll dig into it.

**Comment by kdy1** [comment]
@overlookmotel Can you make a PR if it's good enough?
I can add some patches to the PR if required. (e.g. rust code)

**Comment by overlookmotel** [comment]
@kdy1 Yes I'd be happy to. And thanks for offer of help with the Rust code.

My plan was:

1. Benchmark different serialization frameworks (protobuf, bincode, cap'n proto, rkyv) on a small set of JS syntax and see which works faster.
2. Try to get the winner working for the entire set of AST nodes (I say "try" because I might need some help with that - the errors I was getting with serde/bincode were incomprehensible to me).
3. Apply same to `print()`.
4. Sketch out what new plugin system would look like to take advantage of lazy deserialization.

I'm very busy with day job, so this may all take me some weeks/months, but I'm really interested so will progress it whenever I get a chance.

One question:

In my mind, the ideal end point is:

1. For SWC to have a Babel-compatible plugin system (i.e. SWC can use Babel plugins without modification).
2. For those to run *at minimum* faster than in Babel.
3. Hopefully they run fast enough that there isn't an order of magnitude difference between writing an SWC plugin in JS and in Rust.

i.e. aim of the game is compatibility with existing JS ecosystem, and for writing SWC plugins in JS to be a viable (and still fairly speedy) option rather than "if you don't know Rust, you can't do it".

However, I'm new to the project, so I'm not aware what your vision/priorities for SWC are. Is the above compatible with the direction you want to take?

**Comment by overlookmotel** [comment]
Sorry, one other question: Are the Typescript type definitions in [types.ts](https://github.com/swc-project/swc/blob/master/node-swc/src/types.ts) and Visitor class in [Visitor.ts](https://github.com/swc-project/swc/blob/master/node-swc/src/Visitor.ts) written by hand, or programmatically generated somehow from the Rust type definitions? If the latter, can you point me to the code which does this please?

**Comment by kdy1** [comment]
> However, I'm new to the project, so I'm not aware what your vision/priorities for SWC are. Is the above compatible with the direction you want to take?


Yes, I think compatibility with babel plugin is the most important. 

But I don't expect js plugins to be speedy as rust plugins, because they can't be called in threads.
It's enough I think, as I'm going to provide lots of port of babel plugins by myself.


> Sorry, one other question: Are the Typescript type definitions in types.ts and Visitor class in Visitor.ts written by hand, or programmatically generated somehow from the Rust type definitions? If the latter, can you point me to the code which does this please?

It's written by hand. I actually tried generating it, but I stopped because rust proc macro api is limited.

**Comment by kdy1** [comment]
I absolutely do not mind :)
Thanks!

**Comment by yisar** [comment]
There is another way, that is, we modify the fake AST in the node layer, and then generate the smallest instruction set and pass it to Rust.

This is also a basic strategy of cross thread communication in the past. Its core point is to generate the minimum instruction set, anyway.

**Comment by kdy1** [comment]
@yisar I don't think it will work. Even if it's implemented, it will be very slow becuase such api means we have to cross language boundaries frequently.

**Comment by yisar** [comment]
@kdy1 In fact, there is another another way, that is, If we can find a way so that Rust can communicate directly with V8 and bypass JS, which is also very good. 

I don't know if node can do it, but deno does it.

https://github.com/denoland/serde_v8

However, please note that the structure of deno is very simple. It is usually just a queue containing function name and arguments.

**Comment by Brooooooklyn** [comment]
If using v8 api directly, we will lost ABI-stable which provided by Node-API

**Comment by kdy1** [comment]
I read this again, and I think the performance will be bad again if we fully imlplement the parser.

JIT is very good for this kind of microbenchmarks.

**Comment by kdy1** [comment]
Turns out that swc was serializing json from the js thread.

Need to revisit this after https://github.com/swc-project/swc/pull/3380 is published

**Comment by kdy1** [comment]
```


> @overlookmotel/swc-parse-test@0.0.1 bench
> node benchmark/bench.js

Running "100 lines (1.58 kB)" suite...
Progress: 100%

  swc:
    2 297 ops/s, ±0.52%   | 65.1% slower

  swc (without deserialization):
    6 582 ops/s, ±0.31%   | fastest

  experiment 1 - swc with custom JSON parser:
    3 642 ops/s, ±0.31%   | 44.67% slower

  babel:
    2 015 ops/s, ±4.30%   | slowest, 69.39% slower

Finished 4 cases!
  Fastest: swc (without deserialization)
  Slowest: babel

Saved to: /Users/kdy1/projects/lab/swc-parse-test/benchmark/100 lines.chart.html
Running "1000 lines (17.78 kB)" suite...
Progress: 100%

  swc:
    224 ops/s, ±1.52%   | 62.6% slower

  swc (without deserialization):
    599 ops/s, ±0.32%   | fastest

  experiment 1 - swc with custom JSON parser:
    338 ops/s, ±0.45%   | 43.57% slower

  babel:
    215 ops/s, ±3.27%   | slowest, 64.11% slower

Finished 4 cases!
  Fastest: swc (without deserialization)
  Slowest: babel

Saved to: /Users/kdy1/projects/lab/swc-parse-test/benchmark/1000 lines.chart.html
Running "10000 lines (197.78 kB)" suite...
Progress: 100%

  swc:
    21 ops/s, ±1.64%   | 58.82% slower

  swc (without deserialization):
    51 ops/s, ±2.23%   | fastest

  experiment 1 - swc with custom JSON parser:
    31 ops/s, ±1.58%   | 39.22% slower

  babel:
    5 ops/s, ±88.40%    | slowest, 90.2% slower

Finished 4 cases!
  Fastest: swc (without deserialization)
  Slowest: babel

Saved to: /Users/kdy1/projects/lab/swc-parse-test/benchmark/10000 lines.chart.html

swc-parse-test on  master [!?] is 📦 v0.0.1 via  v17.4.0 via 🦀 v1.59.0-nightly took 1m8s
❯
```

![image](https://user-images.githubusercontent.com/29931815/152703658-d8e454a4-8dff-4ef3-a17e-9be91585eb68.png)




Note: In real world, the perf of swc should be multipled by the number a bit smaller than number of cpu cores. So the swc parser is much faster in real world, and more importantly while swc is parsing javascript thread can do what they want to do.

**Comment by jantimon** [comment]
does that mean a js plugin api could be fast enough?

**Comment by kdy1** [comment]
I don't think so, unless we embed some custom runtime for js which allow running js code in parallel

**Comment by ScottAwesome** [comment]
> I don't think so, unless we embed some custom runtime for js which allow running js code in parallel

@kdy1 would this be a reasonable thing to think about with QuckJS? Small footprint, supports most (if not all) of JavaScript language semantics, and has a really small footprint compared to trying to do something like embedding Node.

https://bellard.org/quickjs/


On the other hand, you could just embed Deno: https://deno.land/manual/embedding_deno

**Comment by kwonoj** [comment]
We probably need separate discussions about plugin-written-in js perf to this, but in my opinion, embedding a whole (or partial) js runtime engine is too heavyweight. Of course that'll resolve some of the issues we are facing but it'll bring lot other potential questions like bytesize, performance of transform itself, also embedding doesn't mean to get away with any serialization cost - swc's host and js runtime boundary still need some sort of serialization, though it's not string-json based one - and ergonomics support as well (i.e majority of users will expect 'js' plugin can be node.js compatible they can use npm pkgs with cjs import).

I know this won't solve all of the issues around plugin especially for the ppl want to write plugin using js, but new experimental plugin support have meaningful progress recently (https://github.com/swc-project/swc/issues/3167) which I hope to roll out first and let users to write plugins.

**Comment by jantimon** [comment]
For me personally the language doesn't matter - however cross compilation for plugin would make it really hard to write, distribute and maintain cross platform plugins.

That's the only reason why I am keep asking for JS..

**Comment by kdy1** [comment]
@jantimon Our wasm-based plugin system is cross-platform.

**Comment by kdy1** [comment]
Although we have to do more work, we only build one `.wasm` file per plugin.

**Comment by jantimon** [comment]
wow that's so cool!
are there already any docs?

**Comment by kdy1** [comment]
Currently, https://github.com/swc-project/swc/tree/b5a5f2b624430c4be1c91541fac662ed022e6284/crates/swc_plugin tihs is the only documentation.
It will be documented properly soon.

**Comment by overlookmotel** [comment]
I *finally* got some time to work on this.

## What I've done

I decided to try using the RKYV-encoded binary representation of AST which is used in the new WASM plugins as the serialization format, and write a deserializer in JS.

https://github.com/overlookmotel/swc/tree/parse-speed2

This is a bit less of a toy than previous experiments, as it's using SWC's actual parser on the Rust side. The JS deserializer only handles a small subset of AST nodes at present, but it's created with a [codegen](https://github.com/overlookmotel/swc/blob/parse-speed2/deserialize/generate.js) so would not take too long to expand to the full set of AST nodes.

Here are the results:

![1000 lines](https://raw.githubusercontent.com/overlookmotel/swc/parse-speed2/bench.png)

## Findings

* `parse()` is speeded up by approx x2.
* Both serialization in Rust and deserialization in JS do have a significant cost.
* Unlike using JSON as serialization format, there is negligible cost passing the data from Rust to JS.
* It's still much slower than Babel!

## Questions

### SWC vs Babel

The relative speed of SWC vs Babel I'm seeing doesn't at all match @kdy1's result [in comment above](https://github.com/swc-project/swc/issues/2175#issuecomment-1030926102).

@kdy1 Could you possibly run this latest [benchmark](https://github.com/overlookmotel/swc/blob/parse-speed2/bench.js) on your machine and see if you see same results as me? I'm wondering if for some reason Babel performs much better on my machine than yours, or if I've compiled SWC incorrectly and hobbled its speed.

I ran the bench on NodeJS v16.14.2, on a 2019 27-inch iMac with 8-core 3.6GHz Intel Core i9, 24GB RAM.

### Deserialization performance

Could anyone with a better knowledge of JS perf than me advise if anything obvious I can do to speed up the deserializer?

The generated deserializer is [here](https://github.com/overlookmotel/swc/blob/parse-speed2/deserialize/index.js) and this is the [codegen](https://github.com/overlookmotel/swc/blob/parse-speed2/deserialize/generate.js).

In particular, it uses the `.readUInt32LE()` method of NodeJS's `Buffer` class a lot. Is there a faster way to get numerical values from a Buffer/ArrayBuffer?

I would be very grateful of any advice.

### Serialization

I'm surprised serialization with RKYV is so costly. Presumably you're seeing the same effect with WASM plugins?

Or is my method of turning RKYV's output into a buffer producing a slow-down?

https://github.com/overlookmotel/swc/blob/fba4e55672eee357eb9724e1a5cfeeef76a546e2/crates/node/src/parse.rs#L265-L267

### Is this benchmark useful?

I am confused why Acorn is so slow in comparison to Babel. I thought Acorn was meant to be faster. Perhaps this indicates input source code in the benchmark (a string of `const x = 123;` statements) is so unrepresentative of real world JS, that this benchmark is not useful?

## Why I care about this

Just to explain why I'm keen for SWC to have a JS plugin system and for it to be *somewhat* fast...

I see two main use cases:

1. Rapid prototyping of plugins, which can later be ported to Rust.
2. Ability for anyone without Rust skills (like me!) to write plugins.

I can also see scenarios like this occurring:

* Bob doesn't know Rust. He writes a plugin in JS and publishes it to NPM.
* Daisy finds the plugin useful, but it's not fast enough for her. She does speak Rust. She ports the plugin to Rust.

I suspect that many people are currently blocked from switching to SWC due to missing support for their favourite Babel plugin. If there are more plugins available (even if written in slow JS), that may boost adoption. That then increases the chance that someone, somewhere has the incentive to port the JS plugins they use to Rust. It could be a virtuous circle.

**Comment by kdy1** [comment]
wasm plugin is slow on first run, because swc should compile wasm to a bytecode.
But it's cached.
I'll run the benchmark soon.

**Comment by kdy1** [comment]
Of course, but `rkyv` was super fast when I profiled it.  (I drunk, sorry if this or previous comment sounds strange)
I cloned your fork, will test soon.

**Comment by overlookmotel** [comment]
Drunk at the keyboard! Disgraceful!

**Comment by kdy1** [comment]
```
➜  js-plugin-bench git:(parse-speed2) ✗ node bench.js
Running "100 lines (1.58 kB)" suite...
Progress: 100%

  swc WASM:
    1 908 ops/s, ±4.56%    | slowest, 89.16% slower

  swc:
    3 204 ops/s, ±0.18%    | 81.8% slower

  swc with buffer serialization and deserialization:
    7 246 ops/s, ±0.50%    | 58.85% slower

  swc with buffer serialization but no deserialization:
    13 243 ops/s, ±0.40%   | 24.79% slower

  swc with buffer serialization but buffer not returned to JS:
    13 749 ops/s, ±0.14%   | 21.91% slower

  swc with no serialization or deserialization:
    17 607 ops/s, ±0.18%   | fastest

  swc with custom JSON deserialization:
    5 114 ops/s, ±0.14%    | 70.95% slower

  swc with JSON serialization but no deserialization:
    9 118 ops/s, ±0.28%    | 48.21% slower

  swc with JSON serialization but JSON not returned to JS:
    9 450 ops/s, ±0.29%    | 46.33% slower

  babel:
    16 347 ops/s, ±0.19%   | 7.16% slower

  acorn:
    14 641 ops/s, ±0.24%   | 16.85% slower

Finished 11 cases!
  Fastest: swc with no serialization or deserialization
  Slowest: swc WASM

Saved to: /Users/kdy1/projects/js-plugin-bench/100 lines.chart.html
Running "1000 lines (17.78 kB)" suite...
Progress: 100%

  swc WASM:
    233 ops/s, ±0.72%     | slowest, 87.91% slower

  swc:
    329 ops/s, ±0.15%     | 82.93% slower

  swc with buffer serialization and deserialization:
    790 ops/s, ±0.63%     | 59% slower

  swc with buffer serialization but no deserialization:
    1 537 ops/s, ±0.23%   | 20.24% slower

  swc with buffer serialization but buffer not returned to JS:
    1 559 ops/s, ±0.12%   | 19.1% slower

  swc with no serialization or deserialization:
    1 927 ops/s, ±0.07%   | fastest

  swc with custom JSON deserialization:
    525 ops/s, ±0.25%     | 72.76% slower

  swc with JSON serialization but no deserialization:
    944 ops/s, ±0.16%     | 51.01% slower

  swc with JSON serialization but JSON not returned to JS:
    1 022 ops/s, ±0.06%   | 46.96% slower

  babel:
    1 743 ops/s, ±0.14%   | 9.55% slower

  acorn:
    583 ops/s, ±0.12%     | 69.75% slower

Finished 11 cases!
  Fastest: swc with no serialization or deserialization
  Slowest: swc WASM

Saved to: /Users/kdy1/projects/js-plugin-bench/1000 lines.chart.html
Running "10000 lines (197.78 kB)" suite...
Progress: 100%

  swc WASM:
    23 ops/s, ±1.57%    | 87.77% slower

  swc:
    30 ops/s, ±0.88%    | 84.04% slower

  swc with buffer serialization and deserialization:
    73 ops/s, ±1.69%    | 61.17% slower

  swc with buffer serialization but no deserialization:
    143 ops/s, ±0.57%   | 23.94% slower

  swc with buffer serialization but buffer not returned to JS:
    144 ops/s, ±0.33%   | 23.4% slower

  swc with no serialization or deserialization:
    188 ops/s, ±0.17%   | fastest

  swc with custom JSON deserialization:
    50 ops/s, ±0.89%    | 73.4% slower

  swc with JSON serialization but no deserialization:
    93 ops/s, ±0.18%    | 50.53% slower

  swc with JSON serialization but JSON not returned to JS:
    101 ops/s, ±0.20%   | 46.28% slower

  babel:
    163 ops/s, ±1.10%   | 13.3% slower

  acorn:
    8 ops/s, ±0.10%     | slowest, 95.74% slower

Finished 11 cases!
  Fastest: swc with no serialization or deserialization
  Slowest: acorn

```


I'm using m1 max pro macbook 2021, 16 inch 64GM ram

**Comment by kdy1** [comment]
Finally wasm-pack is supported on m1 mac

**Comment by kdy1** [comment]
Oh, I looked at code and I found that `praseSyncWasm` is calling `@swc/wasm.parseSync`.
It's not related to wasm plugin. It's wasm-based version of `@swc/core`

**Comment by overlookmotel** [comment]
Thanks for running the benchmark. You and I are seeing roughly the same times now.

Sorry, the thing about WASM was a distraction. What I meant was: You have implemented a binary serialization format for ASTs (for use in WASM plugins). I thought I'd use that same format for a different purpose (without WASM). I only added SWC WASM to the benchmark as I wanted to investigate if much of the time was spent crossing the JS/native boundary, and wondered if WASM would be faster. It's not.

Concerning speed of `rkyv`: According to my benchmark, `rkyv` is only about 35% faster than `serde`. Please see chart above (may need to refresh the page - I've updated it).

That does seem strange. Maybe I'm doing the `rkyv` serialization on the wrong thread or something [here](https://github.com/overlookmotel/swc/blob/fba4e55672eee357eb9724e1a5cfeeef76a546e2/crates/node/src/parse.rs#L265-L267)?

**Comment by overlookmotel** [comment]
For any performance people out there: Flame graph of JS deserializer reveals majority of the time is spent in `Abort.ExtraWide [CODE:BytecodeHandler]`. Google reveals very little on what this is or means. Does anyone have any pointers?

**Comment by kwonoj** [comment]
> Maybe I'm doing the rkyv serialization on the wrong thread or something [here](https://github.com/overlookmotel/swc/blob/fba4e55672eee357eb9724e1a5cfeeef76a546e2/crates/node/src/parse.rs#L265-L267)? (please excuse my ignorance of Rust)

can we rule out copy cost at least? something like this maybe?

```
    let s = Serialized::serialize(&program).unwrap();
    let b: Buffer = Buffer::from(s.as_ref().as_slice());
```

rkyv is not 100% zero-cost: it involves some costs indeed. I don't have strong opinions around if it's our code doing something wrong or there's nature of the AST structures brings some costs around though.

> I wanted to investigate if much of the time was spent crossing the JS/native boundary, and wondered if WASM would be faster. It's not.

I know this is off topic, but swc/wasm still crosses js/native boundary anyway. isn't it?

**Comment by overlookmotel** [comment]
Thanks @kwonoj. I've made that changed. It made a few % difference (well I think so, am seeing a lot of variability across runs).

I've also made some small changes to the JS deserializer which have almost doubled its speed. Overall, now 180% faster than current `swc.parseSync()` (although that might go down once all AST nodes are implemented).

![1000 lines](https://raw.githubusercontent.com/overlookmotel/swc/parse-speed2/bench2.png)

I didn't expect `rkyv` to be zero cost, but I do find it surprising that it's only 35% faster than serde JSON. I feel like this can't be right and I'm doing something wrong.

**Comment by kwonoj** [comment]
>  but I do find it surprising that it's only 35% faster than serde JSON

I feel we may need some isolation to prove where does it comes from. Possible guesses are

- My implementation around rkyv::to_bytes is wrong
- There are overhead to pass over constructed napi buffervalue to v8
- Our AST struct have some inefficiency causes rkyv performs poor for certain operations
- it really only tops ~35% max maybe

I don't have full history but back in the moment when we compared different serialization for the plugin, abi / serde / rkyv, rkyv was notably performs better enough to use it. That makes me think we have some unknown bottlenecks in current code.

**Comment by overlookmotel** [comment]
From what little I know about WASM, I believe that in some circumstances the cost of calling between JS and WASM can be lower than between JS and native.

https://hacks.mozilla.org/2018/10/calls-between-javascript-and-webassembly-are-finally-fast-%F0%9F%8E%89/

I thought it might make a difference in this case as parse is so fast anyway, that the boundary-crossing time could be significant. But sadly not!

**Comment by RiESAEX** [comment]
I guess if we give `rkyv` larger buffer and scratch, it can be more faster, even near to zero cost.
like in this benchmark.
https://github.com/djkoloski/rust_serialization_benchmark/blob/c164d2c4a1e8f44efeebbac8a4efa70ed54bfdf1/src/bench_rkyv.rs#L27-L45

**Comment by kwonoj** [comment]
> cost of calling between JS and WASM can be lower than between JS and native.

The article you shared about is `call`, the part we have to deal in here is `memory`. Since wasm doesn't allow to share memory directly for typed structures (except sharedarraybuffer, which we can't use anyway) we have to _copy_ with serialized structes into the wasm's memory space which is not free & somewhat costly due to wasm's memory space structures.

**Comment by overlookmotel** [comment]
`rkyv` speed: Interesting. But way above my pay grade!

> * There are overhead to pass over constructed napi buffervalue to v8

I thought I'd ruled that one out. See difference between bars 4 and 5 on the chart.

* Bar 4 is serializing to buffer and sending buffer to JS, but then doing nothing with it.
* Bar 5 is serializing to buffer and then doing nothing with it in Rust. Rust returns an empty string to JS.

https://github.com/overlookmotel/swc/blob/parse-speed2/crates/node/src/parse.rs#L309-L312

There's practically no difference between them.

**Comment by overlookmotel** [comment]
It'd be really interesting to see if increasing `rkyv`'s buffer and scratch makes a difference. Would you be willing to try? Or please could you point me in the right direction?

**Comment by kwonoj** [comment]
I think this is interesting topic and we'll potentially leave it as opened for further exploration.

But also meanwhile, I'd like to set some practical goals as well.

- With some of improvements we made, do we think current performance of `parse` is somewhat acceptable?
- If not, what's our baseline goal we want to achieve?

In my opinion I think it is reasonably ok. It really depends on the scale & usecase though. If there are some ecosystem requires to use `parse` only to get AST into v8 and if we want to make it as first-class usecase we may need more digging. But so far in my opinion primary usecase for the swc/core is transform.

For the usecase of parse-plugin-in-js, I'd like to suggest to separate it for now. There are lot of challenges to support it well enough and current primary focus is let user write plugin in wasm directly. I am totally aware lot of users are not familiar with non-js plugin ecosystem and there are some valid suggestions to bootstrapping ecosystem via js-based plugins though.

**Comment by RiESAEX** [comment]
![image](https://user-images.githubusercontent.com/89718456/161323401-4ee8fc0d-b9b3-4340-a3b7-eedb3c3ccbd9.png)
```rust
    pub fn serialize_with_custom<W>(t: &W) -> Result<Serialized, Error>
    where
        W: rkyv::Serialize<rkyv::ser::serializers::AllocSerializer<512>>,
    {
        const BUFFER_LEN: usize = 4096;
        // const SCRATCH_LEN: usize = 512;
        let mut serialize_buffer = AlignedVec::with_capacity(BUFFER_LEN);
        // let mut serialize_scratch = AlignedVec::with_capacity(SCRATCH_LEN);
        let mut serializer = CompositeSerializer::new(
            AlignedSerializer::new(&mut serialize_buffer),
            FallbackScratch::new(HeapScratch::<512>::default(), AllocScratch::default()),
            SharedSerializeMap::default(),
        );
        let field = serializer.into_serializer().into_inner();
        Ok(Serialized {
            field: field.to_owned(),
        })
    }
```
it runs too fast..May be I write some wrong code or give it too many memory?

**Comment by Gregoor** [comment]
A short comment from the sidelines (really enjoy reading along, thanks for thinking out loud!): I’m investigating using swc for my projectional code editor where babel is a tad too slow but I still need babel ast compatibility for prettier.
Probably outside of what this here is intended to solve for, just thought I’d at least bring it up.

I’m hoping for 10x on babel but it looks like an incremental parsing approach might be a better bet for my purposes ultimately.

Anyway, thanks for all the hard detailed work!

**Comment by overlookmotel** [comment]
@kwonoj On the question of how fast is fast enough:

I can see 3 ways to approach this:

1. Faster than Babel.
2. Fast enough to make JS plugins fast enough.
3. As fast as possible.

### Faster than Babel

Personally I think this would be a legitimate aim. But it's not clear right now if it is achievable.

Judging by benchmarks above, SWC parse + RKYV serialization (without any deserialization) takes almost as long as Babel takes in total. Once deserialization is added (no matter how efficient it is), it's hard to believe it'll not push performance below Babel.

However...

Due to the poor performance of Acorn in benchmark above, I've become suspicious that the input in my benchmark is too simple (it's only a series of `const x = 1;` statements), and the result may not be a real measure. I'm working on implementing all AST nodes in the deserializer so will then be able to benchmark on some real JS code. That should give us a much better idea of where we're at. That work is nearly done - should have it finished in a week or two.

Secondly, RKYV's serialization (as far as I can see from experiments above) is surprisingly slow. If that continues to be the case when it's run on a more representative sample of code, that'll be the place to look for gains.

I suspect this can't be right. From my (limited) understanding, what RKYV is doing isn't hugely complex, so in theory at least I'd expect it to be faster.

But if RKYV really is as slow as it looks, I think that's going to cause you performance problems with WASM plugins. As I understand it, WASM plugins require two rounds of serialization/deserialization - once on the way in to the plugin, and a second round on the way out. If the user is using 3 plugins, that'll result in 6 rounds of serialization/deserialization. This could negate a lot of the benefit of near-native performance in the plugin itself.

My point here is not to be down on WASM plugins. I'm not! What I'm trying to say is that I imagine that if it does turn out like this, speeding up serialization will become your focus, for the benefit of WASM plugins.

Either way, if AST serialization gets faster, that would also benefit `parse()`, and it *might* be enough to make SWC's parser competitive with Babel's.

So, the jury is out for now on whether SWC can beat Babel or not.

### Fast enough to make JS plugins fast enough

This is the real prize in my mind. However, I totally understand that this is not your focus, and you'd like to keep it separate, so I'll skip over that.

### As fast as possible

In the meantime, I'm just heading for "as fast as possible".

Unless SWC intends to remove `parse()` entirely, I assume you wouldn't object to a PR that makes an existing API faster than it is at present?

Again, I understand that this is not anyone's focus, and I don't expect much input. I hope you don't mind if I continue plugging away at it when I can find time.

I'll pop up again when I have the deserializer working on the full set of AST nodes, and have a measure of how fast it is on real code. I hope that'll give us something more concrete to base this discussion on.

**Comment by overlookmotel** [comment]
@Gregoor I'm sorry to say that 10x Babel's speed is *never* going to be possible. ASTs are huge nests of many many objects. Creating all those objects in JS takes time, regardless if the actual parser runs at the speed of light.

If you're looking for that kind of performance, you'd need to go native. Or, as you say, use incremental parsing.

**Comment by overlookmotel** [comment]
@RiESAEX Thanks loads for experimenting with RKYV's settings. Disappointing it's not produced a dramatic gain (I was pretty excited when I saw your first results, which turned out to be a mirage), but helpful to know nonetheless.

**Comment by kwonoj** [comment]
Not a high-level perspective, just random notes:

- `(it's only a series of const x = 1; statements)`
I agree benchmark may need to be more realistic to resemble some of actual code paths being used in the applications. Especially, this effectively tests only specific path of the parser in both lands.

- Secondly, RKYV's serialization (as far as I can see from experiments above) is surprisingly slow.

This is somewhat interesting to figure out where it actually comes from. When we did initial investigation, it was definitely faster than json-based serialization (serde_*), and was closer enough to native abi (abi_stable). Could be multiple reasons and if it's something comes from the way swc utilizes it in certain way such as causing pressure to buffer alloc / else it may need to be fixed. 

In a plugin's transform, still it _looks_ like reasonably acceptable. I really can't tell where this difference comes. If anyone have isolated repro cases and run trace over those code paths it maybe helpful.

- Fast enough to make JS plugins fast enough

As said, yes, this is not a primary goal to achieve. To JS plugin to work, there must be a step to construct passed ast into JSON object at some point. That's something I doubt we can achieve reasonable performance unless there's huge breakthrough. Also there are other blockers, logistics issues to make js plugins as first class citizen for the swc. I'm not saying hard no, but this isn't something we are actively aim to focus.

-  I assume you wouldn't object to a PR that makes an existing API faster than it is at present

We are welcome any feasible improvement for the codebases. We do not have any immediate plan to deprecate `parse` anytime soon.

Last random note is now this is mostly becomes discussion around transfer (or copying) ast struct between two different context. Native parse perf should matter, but pretty much most of last discussions are about the cost to pass we have to pay for eventually just matter of the amount. Even in the JS-only context if you have to deal with similar (i.e worker, or ipc, or anything else) it becomes interesting challenges as either plain cloning or structured cloning, or `serializing` js object into buffer to transfer between since all of those have visible costs.

**Comment by overlookmotel** [comment]
@kwonoj Thanks for swift reply.

OK great, I'll keep plugging away at implementing the full deserializer, so we can get some more "real world" numbers.

We may also find that the unintuitive "slow RKYV" effect disappears at that point. Let's see...

You're right, this is all about the transfer between "worlds", and nothing to do with parsing itself! It's frustrating this is such a pain point, but yeah the fact that it's a common problem does make it interesting.

And just briefly:

> To JS plugin to work, there must be a step to construct passed ast into JSON object at some point.

This is actually where I disagree with you. Using a binary representation as the serialization medium opens the door to convert to JSON *only the specific sections* of the AST which a plugin needs to visit - which is typically only a small part of the whole AST. I'm not 100% sure this will work, but I'd like to try. *If* it does work, that could be the breakthrough you mention.

I hear very clearly your position that this is not a focus for SWC. But thanks for not ruling it out entirely.

I think best I just shut up about it for now, do some experiments, and come back when/if I have a proof of concept.

Anyway, `parse()` is first...

**Comment by kwonoj** [comment]
> Using a binary representation as the serialization medium opens the door to convert to JSON only the specific sections of the AST which a plugin needs to visit 

This only works if binary representation of given raw buffer is 1. have strict format allows partial read gauranteed specific position of struct 2. first is gauranteed across any platform swc supports (including wasm) and probably additional constraints.

What you mention is essentially a `deserialization`, difference is you're proposing a new deserialization mechanism which supports `partial` deserialization of given bytes. I am not an expert but some of serialization have partial support to so called `lazy` deserialization for some portion of given bytes but it requires to copy over whole bytes still + not generally available for any arbitrary portion of given schema. I doubt this is something easily doable across all the platforms.

Moreover, that doesn't mean you can't get rid of partially deserialized byte into `js object`, in the end js code can only work with js object, not a raw buffer for the AST. Partial may give gain to reduce amount of constructing js object, but that doesn't mean you can get rid of all of them.

**Comment by golopot** [comment]
I wrote a script to find out the cost of parse time versus json serialization/deserialization time. Using acorn parser on a realistic Javascript file. It turns out json serialization/deserialization time accounts for about 20% ~ 30%  of total time.
 
```js
async function main() {
  const { parse } = await import(
    "https://unpkg.com/acorn@8.7.0/dist/acorn.mjs"
  );
  const source = await fetch(
    "https://unpkg.com/react@18.0.0/cjs/react.development.js"
  ).then((r) => r.text());

  console.log("source length: " + source.length);
  let start = performance.now();
  let ast = parse(source, { ecmaVersion: 2020 });
  let end = performance.now();
  console.log(`parse time: ${end - start} ms`);

  start = performance.now();
  let o = JSON.stringify(ast);
  end = performance.now();
  console.log(`JSON.stringify time: ${end - start} ms`);

  start = performance.now();
  ast = JSON.parse(o);
  end = performance.now();
  console.log(`JSON.parse time: ${end - start} ms`);
}

main();

```

**output:**
```
"source length: 87561"
"parse time: 27.19999998807907 ms"
"JSON.stringify time: 4.100000008940697 ms"
"JSON.parse time: 3.9000000059604645 ms"
```

**demo link:**
https://jsbin.com/zusewab/edit?js,console

**Comment by Gregoor** [comment]
> @Gregoor I'm sorry to say that 10x Babel's speed is _never_ going to be possible. ASTs are huge nests of many many objects. Creating all those objects in JS takes time, even if the actual parser runs at the speed of light.
> 
> If you're looking for that kind of performance, you'd need to go native. Or, as you say, use incremental parsing.
> 
> Sorry to be a party-pooper!

I appreciate it, as displayed I do not know about the orders of magnitude involved here, so this is a good reality check to get. I’ll chase after incremental parser then!

still lotta love for swc though <3

**Comment by kwonoj** [comment]
I'd like to suggest to change title of this issue to narrow down to specific issue (serialization / deserialization + transfer cost) and if there's an internal parse perf issue file a new issue separately.

I'll do it by around end of the day on my end, roughly 12hrs later if there are no objections.

**Comment by overlookmotel** [comment]
@kwonoj No objection from me. You're right, that has turned out to be the subject here.

**Comment by kwonoj** [comment]
😓 I forgot I am member but does not have some write permissions. I'll leave it to @kdy1 to change title accordingly.

**Comment by kdy1** [comment]
I changed the role
Seems like I made some mistake while configuring @swc-bot

**Comment by overlookmotel** [comment]
PS My intent is not to go nuts trying to squeeze every last drop of performance and posting about this forever. I feel like it's moving towards putting together a PR pretty soon.

However, it does feel like there's some low-hanging fruit which could yield significant gains, so I'd be keen to try to resolve those things if possible before making a PR.

**Comment by dannymcgee** [comment]
> ```rust
> let serialized = Serialized::serialize(&program).convert_err()?;
> let slice = serialized.as_ref().as_slice();
> let buffer = Buffer::from(slice);
> Ok(buffer)
> ```
> [...]
> Presumably this is just a mem copy operation, so it seems odd it's so slow. (the code is [here](https://github.com/overlookmotel/swc/blob/parse-speed2/crates/node/src/parse.rs#L265-L269))

Sorry for the interruption — I just read through most of this thread and wanted to say it's delightful to see how fast you went from "I don't know any Rust D:" to [practicing the Dark Arts](https://github.com/overlookmotel/swc/commit/310362196217a858af72f755c489b95a5b9c5f5a#diff-5df32545e1b3a38807877c6dd2a04c54c0304b0018a1becfc27a69c1ac41984cR314). (I actually came to suggest exactly that, but then noticed your link was broken and saw that you had already done it in the meantime.)

I'm investigating swc for an imminent work project, and mostly what I was wondering is whether I can start using the released/documented version in a couple of weeks and upgrade relatively seamlessly down the road, or if I'll need to eventually rewrite my plugins in Rust. It sounds like the JS API will stay put, but any perf improvements over the status quo would be dependent on the work you're doing @overlookmotel (and aren't a big priority for the swc team in any event)? If there's anything I can do to help, please feel free to ping me!

**Comment by overlookmotel** [comment]
@dannymcgee Sorry for slow reply. Day job (which has nothing to do with compilers, or indeed programming) went nuts for a while, and then I got Covid!

Well I still don't think I can say I know Rust! I'm not so much "practicing the dark arts" as flinging darkness at the compiler until something works. I am really enjoying getting to know Rust, but it's a slow process.

My feeling is that in the end JS plugins *can* be made fast enough to be practical. However, the work discussed in this thread is only phase 1. There's potential to make them *much* faster than the above by implementing partial/lazy serialization/deserialization, and I think this could get them to the point of being almost as fast as WASM plugins in a lot of cases. However, there are some complications due to the fact that AST nodes cannot be repeated (i.e used more than once) in Rust, whereas they can in JS. So that needs to be handled.

So overall, I do think it's possible, but I only have so much time to work on it. It could take a *long* time before it comes to fruition.

If you do have some time to lend a hand, any help with the two points above would be massively appreciated. Implementing TypeScript support is now the main blocker on completing this "phase 1" work.

**Comment by kwonoj** [comment]
I'm not able to dig deeper into the findings & suggestions for now, just want to double clarify one thing:

This issue is strictly trying to pursue serialization / deserialization + transfer performance improvement and would love to adapt any changes that can improve those.

But this doesn't mean having this automatically SWC core team will try to pursue to enable JS plugin as first class citizen. Performance / ergonomics around JS plugin have a lot more aspects to consider, which we made a very careful and deliberate decision to pursue wasm plugin instead. I'm not going to say there's no future reevaluation at all or something, but also we don't have answers to a lot of questions if we want to revisit this. 

In short, I'd expect this issue will be the place for improving existing node bindings interfaces like `transform*, parse*, etcs`. It'll be a huge win for the SWC itself for its performance numbers. Other things may need to be discussed separately once we feel this is achieved.

**Comment by overlookmotel** [comment]
@kwonoj Thanks for your swift reply. I totally understand your position and have largely avoided discussion of JS plugins for this reason. I only wanted to respond to @dannymcgee's question.

My intent is to finish up this work and then open a new issue about JS plugins. But the impression I've got is that the best way to try to convince you and the other maintainers that it's even worth considering is to demonstrate some halfway-decent performance, since that was the main objection originally. So that's what I'm working towards.

Just to repeat: I'm very clear that the maintainers don't see that at all as a priority, and consequently I don't expect any help. I hope you don't mind me communicating with others who have the same interest though.

**Comment by kwonoj** [comment]
@overlookmotel Thanks for sharing context and understanding.

> My intent is to finish up this work and then open a new issue about JS plugins

My impression is before thinking about any js plugin and other new discussion, it'd be great if there's working PR against to existing interfaces (`transform_*, parse_*,...`) which immediately can benefit from this for everyone.

> I hope you don't mind me communicating with others who have the same interest though.

Yes sure, that's the whole purpose of keeping opening this issue for the tracking purpose.

**Comment by overlookmotel** [comment]
> My impression is before thinking about any js plugin and other new discussion, it'd be great if there's working PR against to existing interfaces (`transform_*, parse_*,...`) which immediately can benefit from this for everyone.

Absolutely.

One question: If I get to the point where everything is done except TypeScript support, would you accept a PR for a more performant `parse*`, `print*` and `transform*` which falls back to previous slower JSON-based implementation when input is TS?

As you probably gathered from the above, I have very little appetite to implement TypeScript support, and part of me feels it would feel fair enough on an open source project to leave that as a follow-on for someone else who does have the appetite.

I know that wouldn't be entirely ideal, but otherwise it's going to be the blocker for all the rest - could result in it taking many months to get to the PR stage, rather than weeks.

**Comment by kwonoj** [comment]
> If I get to the point where everything is done except TypeScript support, would you accept a PR for a more performant parse*, print* and transform* which falls back to previous slower JSON-based implementation when input is TS?

I really can't say without having details, but it sounds like a possible path if there aren't regressions (by properly falling back to current execution path) & if we can replace into new implementation entirely in a reasonable timeframe. It is not uncommon for us to have several staged PR to make a new feature / refactoring to land.

**Comment by natew** [comment]
@overlookmotel just found out about this and want to say one thing:

Thank you 🙇

This is hugely important work.

For reference I built [Tamagui](https://tamagui.dev) which makes extensive use of babel to do all sorts of optimizations, but that has so much logic all over. The flexibility of JS is imperative to being able to support those features, it's a big technical advantage for shipping and there's still a bunch more that could be done.

I had brought this idea up with babel folks but didn't get positive response, so very glad to see maybe I can port to SWC in the future.

**Comment by kdy1** [comment]
JS plugin will be dropped in v2

**Comment by overlookmotel** [comment]
@kdy1 Can I ask if you could please re-open this?

This issue relates to speeding up the interface for existing APIs - `parse`, `print` etc - NOT to JS plugins. Everyone is jumping on the JS plugins angle (including me) but that's not actually the subject of this issue.

I've done a huge amount of work on this to date and it's close to the finish line. True, it has taken me a very long time, but please bear in mind that (a) I have had to learn Rust from nothing and (b) my day job has nothing at all to do with coding, so I can only work on it in my spare time which, due to the demands of my job, is very minimal for large chunks of the year.

The implementation I have is working, and what I've been (slowly) working on more recently is making the codegen for the JS-side serializer/deserializer be generated from the Rust types automatically, so that it's future-proof for any changes to the Rust code. My intention with that is to ensure it's not a maintenance burden for you and the other maintainers. I really am trying to do this the right way.

As the benchmarks above show, the approach I've taken makes significant performance improvements, and I would very much appreciate the opportunity to get this over the finish line.

**Comment by natew** [comment]
@kdy1 could we get some discussion on why the JS plugin being dropped / have a chance to discuss it somewhere?

**Comment by kwonoj** [comment]
There were multiple discussions regarding plugins at  https://github.com/swc-project/swc/discussions , though it is not a single umbrella discussion so you may need some archeology to find some details.

In short summary, things like performances including serialization and runtime both, supporting various platforms like native cli / wasm host, ergonomics concerns of ast visitor that cannot replicate rust behavior 100%, ecosystem fragmentations. There is no such things like 100% impossible so we may revisit this one day depends on the circumstances, but for now we are focusing to improve current wasm based pluing improvements.

**Comment by kwonoj** [comment]
For the cleanup purpose, I'm going to close this: not because of we will not trying to adapt suggestions in this issue, more of actionable perspective. Currently this issue does not have clear actionable item on @swc/core side and it's hard to keep the issue as opened indefinitely. If there are outcomes from this investigation as a PR, happy to accept it. Or either, if there are an actionable suggestions instead - please file an RFC.
